File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.py
def solution(s):
    answer = True
    
    dict_parentheses = {
        "(": True,
        ")": False,
    }
    
    stack_parentheses = []
    
    for item in s:
        if dict_parentheses[item]:
            stack_parentheses.append(item)
        elif not stack_parentheses:
            return False
        else:
            stack_parentheses.pop()
        
    if stack_parentheses:
        return False

    return True

File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.cpp
#include<string>
#include <stack>

using namespace std;

bool solution(string s)
{
    stack<char> stack_temp;
    
    for (char c : s) {
        if (c == '(' || c == '{') {
            stack_temp.push(c);
        } else {
          if (stack_temp.empty()) return false;
            
          if (c == ')' && stack_temp.top() == '(') {
            stack_temp.pop();
          } else if (c == '}' && stack_temp.top() == '{') {
            stack_temp.pop();
          } else {
            return false;
          }
        }
    }

    return stack_temp.empty();
}